### ISSUE TITLE
Support state data members

### ISSUE THREAD

**Issue by erikzenker** [erikzenker/hsm#43]
The boost msm supports state data members. Which means that your states can have members which you can access on events and actions.

E.g.: 
```c++
    struct action {
        template <class FSM, class EVT, class SourceState, class TargetState>
        void operator()(EVT const& e, FSM& fsm, SourceState& source, TargetState& target)
        {
           if(target.hasMember){
             std::cout << "this state has a member" << std::endl;
           }
        }
    };
```

**Comment by erikzenker** [comment]
It is rather complex to implement this feature. A lot of things need to be changed under the hood.
Thus, it does make sense to split it up into smaller pieces:

1. States need to be encapsulated into `state<MyState>` classes. So that they do not need to be constructed within the transition table
2. `make_transition_table` need to be `static constexpr`. Because I need to call this function without instantiating the class

**Comment by erikzenker** [comment]
I created a wip branch `feature/erikzenker/state-data-members`

**Comment by erikzenker** [comment]
Encapsulating states in `state<MyState>` was merged with #68

**Comment by erikzenker** [comment]
State member support was added with #70
